File: 정종길/프로그래머스/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()
    rocks.append(distance)
    left = 1
    right = distance
    
    while left <= right:
        delete = 0
        mid = (left + right) // 2
        prev_rock = 0
        
        for rock in rocks:
            dist = rock - prev_rock
            
            if dist < mid:
                delete += 1
            else:
                prev_rock = rock
            
            if delete > n:
                break
        
        if delete > n:
            right = mid - 1
        
        else:
            answer = mid
            left = mid + 1
    
    return answer

File: 정종길/프로그래머스/기타/징검다리.java
import java.util.*;
import java.util.stream.Collectors; 

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;
        int start = 0;
        int end = distance;
        int delete, prevRock;
        
        List<Integer> wrRocks = Arrays.stream(rocks)
            .boxed()
            .collect(Collectors.toList());

        Collections.sort(wrRocks);
        wrRocks.add(distance);

        while(start <= end){
            
            int mid = (start + end) / 2;
            delete = 0;
            prevRock = 0;
            
            for(int rock : wrRocks){
                if((rock - prevRock) < mid){
                    delete += 1;
                }else{
                    prevRock = rock;
                }
                
                if(delete > n){
                    break;
                }
            }
            
            if (delete > n){
                end = mid - 1;
                
            } else {
                start = mid + 1;
                answer = mid;
                
            }
        }

        return answer;
    }
}